File: speedtest/ts/speedtest.ts
// linear x 152,109,796 ops/sec ±0.30% (97 runs sampled)
export function findLinear(needle: number, haystack: number[]): number | null {
  let i = 0;
  const n = haystack.length;

  while (i < n - 1 && haystack[i] != needle) {
    i++;
  }

  return haystack[i] == needle ? i : null;
}

// fakeLast x 150,695,846 ops/sec ±0.33% (94 runs sampled)
export function findFakeLast(
  needle: number,
  haystack: number[]
): number | null {
  let i = 0;
  haystack.push(needle);

  while (haystack[i] != needle) {
    i++;
  }

  haystack.pop();
  return i != haystack.length - 1 ? i : null;
}

// binary x 150,779,112 ops/sec ±0.16% (97 runs sampled)
export function findBinary(needle: number, haystack: number[]): number | null {
  let i = 0;
  let j = haystack.length - 1;
  let k = Math.floor(j / 2);

  while (haystack[k] != needle && i < j) {
    if (needle > haystack[k]) {
      i = k + 1;
    } else {
      j = k - 1;
    }
    k = Math.floor((i + j) / 2);
  }

  return haystack[k] == needle ? k : null;
}

File: speedtest/go/speedtest.go
package speedtest

// 4.587 ns/op
func findLinear(needle int, haystack []int) (int, bool) {
	n := len(haystack)
	for i := 0; i < n-1; i++ {
		if haystack[i] == needle {
			return i, true
		}
	}
	return 0, false
}

// 4.701 ns/op
func findFakeLast(needle int, haystack []int) (int, bool) {
	i := 0
	haystack[len(haystack) - 1] = needle

	for ; haystack[i] != needle; i++ {
	}

	if i != len(haystack) - 1 {
		return i, true
	}
	return 0, false
}

// 3.286 ns/op
func findBinary(needle int, haystack []int) (int, bool) {
	i := 0
	j := len(haystack) - 1
	k := j / 2

	for ; haystack[k] != needle && i < j; k = (i + j) / 2 {
		if needle > haystack[k] {
			i = k + 1
		} else {
			j = k - 1
		}
	}

	if haystack[k] == needle {
		return k, true
	} else {
		return 0, false
	}
}
